CustomItems/Items/C4Charge.cs

```diff
@@ -0,0 +1,227 @@
+        {
+            Instance = this;
+
+            ServerEvent.WaitingForPlayers += OnWaitingForPlayers;
+
+            PlayerEvent.ThrowingGrenade += OnThrowing;
+            PlayerEvent.Destroying += OnDestroying;
+            PlayerEvent.Died += OnDied;
+
+            MapEvent.ExplodingGrenade += OnExploding;
+
+            base.SubscribeEvents();
+        }
+
+        /// <inheritdoc/>
+        protected override void UnsubscribeEvents()
+        {
+            Instance = null;
+
+            ServerEvent.WaitingForPlayers -= OnWaitingForPlayers;
+
+            PlayerEvent.ThrowingGrenade -= OnThrowing;
+            PlayerEvent.Destroying -= OnDestroying;
+            PlayerEvent.Died -= OnDied;
+
+            MapEvent.ExplodingGrenade += OnExploding;
+
+            base.UnsubscribeEvents();
+        }
+
+        /// <inheritdoc/>
+        protected override void OnWaitingForPlayers()
+        {
+            PlacedCharges.Clear();
+
+            base.OnWaitingForPlayers();
+        }
+
+        /// <inheritdoc/>
+        protected override void OnThrowing(ThrowingGrenadeEventArgs ev)
+        {
+            if (Check(ev.Player.CurrentItem))
+            {
+                ev.IsAllowed = false;
+
+                ev.Player.RemoveItem(ev.Player.CurrentItem);
+
+                var c4 = Spawn(ev.Player.CameraTransform.position, (ev.Player.Rotation * ThrowMultiplier) + (Vector3.up * 1.5f), FuseTime);
+
+                c4.NetworkthrowerGameObject = ev.Player.GameObject;
+                c4.NetworkthrowerTeam = ev.Player.Team;
+
+                if (!PlacedCharges.ContainsKey(c4))
+                    PlacedCharges.Add(c4, ev.Player);
+            }
+
+            base.OnThrowing(ev);
+        }
+
+        private void OnDestroying(DestroyingEventArgs ev)
+        {
+            foreach (var charge in PlacedCharges.ToList())
+            {
+                if (charge.Value == ev.Player)
+                {
+                    if (DropChargesOnDeath)
+                    {
+                        TrySpawn((int)Id, charge.Key.transform.position);
+                    }
+
+                    NetworkServer.Destroy(charge.Key.gameObject);
+                    PlacedCharges.Remove(charge.Key);
+                }
+            }
+        }
+
+        private void OnDied(DiedEventArgs ev)
+        {
+            foreach (var charge in PlacedCharges.ToList())
+            {
+                if (charge.Value == ev.Target)
+                {
+                    if (DropChargesOnDeath)
+                    {
+                        TrySpawn((int)Id, charge.Key.transform.position);
+                    }
+
+                    NetworkServer.Destroy(charge.Key.gameObject);
+                    PlacedCharges.Remove(charge.Key);
+                }
+            }
+        }
+
+        private void OnExploding(ExplodingGrenadeEventArgs ev)
+        {
+            if (PlacedCharges.ContainsKey(ev.Grenade.GetComponent<Grenade>()))
```

Comment:
`
if (ev.Grenade.TryGetComponent(out Grenade grenade)
{
    PlacedCharges.Remove(grenade);
}
`

Reply:
You shouldn't be calling GetComponent<> multiple times when you don't have to.

Also is there not a CustomGrenade virtual method for Exploding?

Reply:
Fixed component but I don't see the virtual method for grenade explode
![image](https://user-images.githubusercontent.com/75911188/110222402-2988df80-7ed2-11eb-9cf5-6db9126da9c7.png)

Reply:
You're right there isn't, I added it in 2.8.0-rc.1 but I'll handle updating the grenade items to use it.